infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/IngressHostExternalServerExposer.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server;
+
+import static java.lang.Integer.parseInt;
+import static java.util.stream.Collectors.toMap;
+
+import io.fabric8.kubernetes.api.model.ServicePort;
+import io.fabric8.kubernetes.api.model.extensions.Ingress;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides a host-based strategy for exposing service ports outside the cluster using Ingress
+ *
+ * <p>This strategy uses different Ingress host entries <br>
+ * Each external server is exposed with a unique subdomain of CHE_DOMAIN.
+ *
+ * @author Sergii Leshchenko
+ * @author Guy Daich
+ */
+public class IngressHostExternalServerExposer
+    implements ExternalServerExposerStrategy<KubernetesEnvironment> {
+
+  public static final String HOST_STRATEGY = "host";
+  private final String domain;
+  private final Map<String, String> ingressAnnotations;
+  private static final Logger LOG = LoggerFactory.getLogger(IngressHostExternalServerExposer.class);
+
+  @Inject
+  public IngressHostExternalServerExposer(
+      @Named("infra.kubernetes.ingress.annotations") Map<String, String> ingressAnnotations,
+      @Named("che.domain") String domain) {
```

Comment:
Can you elaborate if `che.domain` is a new variable? Does it differ from `che.host`?

Reply:
In host-based routing, we assume that che-master and all workspaces are accessed through different sub-domains of the same domain. For example:
* Che-server would be available at this address: `master.che.my-k8s.com`
* The workspace agent of a specific workspace would be available at this address: `server-4401-serverofflzacl-dev-machine.che.my-k8s.com`
* The user app for the same workspace would be available at this address: `server-8080-serverofflzacl-dev-machine.che.my-k8s.com`
* Other workspaces and corresponding agents

In the example above **`che.host`** is `master.che.my-k8s.com`, whereas **`che.domain`** is `che.my-k8s.com`.

We chose this approach for several reasons:
1. In certain landscapes you might want to serve more than one che-master. This is especially true in continuous integration scenarios, staging environments, etc. In this case `che.domain` could be `che-test.my-k8s.com`, `che-prod.my-k8s.com` and so on.
2. For TLS, we would want to issue a single wildcard certificate for each Che deployment. E.g. `*.che.my-k8s.com`.

We could consider *not* setting the CHE_HOST environment variable upon deployment of che-master, and instead setting it programmatically based on CHE_DOMAIN, but we don't know what effect this would have on the Docker- and OpenShift-based Che infrastructures.

Reply:
I see. Thank you for the detailed explanation! 
From what I see it is used by k8s infra only. So shouldn't we rename it to something k8s infra specific, e.g. `che.infra.kubernetes.che_domain`?

Reply:
I understand the reason why you decided to add it. But can you change the name to be compliant with the naming of other properties? For example `che.infra.kubernetes.ingress.domain`.

Reply:
@garagatyi - agreed. I'll fix it.